### PR TITLE
Add the initial implementation

### PR DESCRIPTION
**Pull request by aashikam**
## Purpose
$title

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,57 @@
...
+[![GitHub Last Commit](https://img.shields.io/github/last-commit/ballerina-platform/module-ballerinax-aws.redshift.svg)](https://github.com/ballerina-platform/module-ballerinax-aws.redshift/commits/main)
+[![GraalVM Check](https://github.com/ballerina-platform/module-ballerinax-aws.redshift/actions/workflows/build-with-bal-test-native.yml/badge.svg)](https://github.com/ballerina-platform/module-ballerinax-aws.redshift/actions/workflows/build-with-bal-test-native.yml)
+[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
+
+The AWS Redshift connector for [Ballerina](https://ballerina.io/) facilitates seamless interaction with Amazon Redshift data warehouses, 
+enabling programmatic access to various applications, databases, APIs, and services through the Java Database 
+Connectivity (JDBC) API using Ballerina. This connector empowers users to execute a diverse set of standard Data 
+Definition Language (DDL) commands, SQL commands, and SQL functions for efficient querying of data sources within the 
+Amazon Redshift environment. Detailed reference.
```
Comment (ThisaruGuruge) [inline README.md]: Let's not use line breaks inside the paragraphs. Some tools renders these differently when we have line breaks.

File: ballerina/build.gradle [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+description = 'Ballerina - Snowflake Ballerina Connector'
```
Comment (ThisaruGuruge) [inline ballerina/build.gradle]:
```suggestion
description = 'Ballerina - Redshift Ballerina Connector'
```

File: ballerina/build.gradle [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+def tomlVersion = stripBallerinaExtensionVersion("${project.version}")
```
Comment (ThisaruGuruge) [inline ballerina/build.gradle]: Let's use the Ballerina Gradle plugin here.

File: gradle.properties [hunk 4]
```diff
@@ -0,0 +1,13 @@
+group=io.ballerinax
```
Comment (ThisaruGuruge) [inline gradle.properties]: Let's use this:
```suggestion
group=io.ballerina.lib
```

File: gradle.properties [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+puppycrawlCheckstyleVersion=10.12.1
+checkstyleToolVersion=10.12.1
+githubSpotbugsVersion=5.0.14
+githubJohnrengelmanShadowVersion=8.1.1
+underCouchDownloadVersion=5.4.0
+researchgateReleaseVersion=2.8.0
+testngVersion=7.6.1
+
+ballerinaLangVersion=2201.8.0-20230830-220400-8a7556d8
+stdlibSqlVersion=1.11.0
```
Comment (ThisaruGuruge) [inline gradle.properties:10]: Let's not use the Ballerina library dependencies here. It should be handled by the Ballerina build. Check other connector (gmail) build scripts.

File: issue_template.md
Comment (ThisaruGuruge) [inline issue_template.md:1]: We don't need issue template in this repo.

File: native/build.gradle [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+plugins {
+    id 'java-library'
```
Comment (ThisaruGuruge) [inline native/build.gradle]:
```suggestion
    id 'java'
```

File: native/build.gradle [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    id 'com.github.spotbugs'
+    id 'checkstyle'
+    id 'jacoco'
```
Comment (ThisaruGuruge) [inline native/build.gradle]: Do we need jacoco here? We don't have any Java tests, right?

File: native/build.gradle [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+description = 'Ballerina - Redshift Native'
+
+configurations {
+    jacocoRuntime
+}
+
+dependencies {
+    jacocoRuntime "org.jacoco:org.jacoco.agent:${jacoco.toolVersion}:runtime"
```
Comment (ThisaruGuruge) [inline native/build.gradle]: We can remove this as well.

File: native/src/main/java/io/ballerina/stdlib/java/jdbc/Constants.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.values.BString;
+
+/**
+ * Constants for AWS Redshift client.
+ *
+ * @since 1.0.0
+ */
+public final class Constants {
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/stdlib/java/jdbc/Constants.java]: Let's add a private constructor here.

File: pull_request_template.md
Comment (ThisaruGuruge) [inline pull_request_template.md:1]: Let's use the PR template from the Ballerina library modules.

File: build.gradle [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+import org.apache.tools.ant.taskdefs.condition.Os
+
+plugins {
+    id "com.github.spotbugs" version "${githubSpotbugsVersion}"
+    id "com.github.johnrengelman.shadow" version "${githubJohnrengelmanShadowVersion}"
+    id "de.undercouch.download" version "${underCouchDownloadVersion}"
+    id "net.researchgate.release" version "${researchgateReleaseVersion}"
```
Comment (ThisaruGuruge) [inline build.gradle]: Let's move the plugin management stuff to the settings.gradle file. Refer the gmail connector repo.

File: ballerina/client.bal [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    *sql:Client;
+
+    public isolated function init(string url, string user, string password,
+        Options? options = (), sql:ConnectionPool? connectionPool = ()) returns sql:Error? {
+        ClientConfiguration clientConf = {
+            url: url,
+            user: user,
+            password: password,
+            options: options,
+            connectionPool: connectionPool
```
Comment (ThisaruGuruge) [inline ballerina/client.bal]: We can use the shorthand notation here.
```suggestion
            URL,
            user,
            password,
            options,
            connectionPool
```

File: ballerina/client.bal [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+import ballerina/jballerina.java;
+import ballerina/sql;
+
+@display {label: "Redshift", iconPath: "icon.png"}
+public isolated client class Client {
+    *sql:Client;
+
+    public isolated function init(string url, string user, string password,
+        Options? options = (), sql:ConnectionPool? connectionPool = ()) returns sql:Error? {
+        ClientConfiguration clientConf = {
```
Comment (ThisaruGuruge) [inline ballerina/client.bal:36]: Let's add a new line to improve readability. I wondering whether the formatting is correct here. Can you check with the formatter?

```suggestion

        ClientConfiguration clientConf = {
```

File: build.gradle [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+description = 'Ballerina - AWS Redshift'
+
+allprojects {
+    group = project.group
+    version = project.version
+
+    apply plugin: 'jacoco'
```
Comment (ThisaruGuruge) [inline build.gradle]: We don't need jacoco here as well.

File: build-config/resources/Ballerina.toml [hunk 14]
```diff
@@ -0,0 +1,25 @@
+[package]
+org = "ballerinax"
+name = "aws.redshift"
+version = "@toml.version@"
+authors = ["Ballerina"]
+keywords = ["Data Warehouse", "Columnar Storage", "Cost/Paid"]
```
Comment (ThisaruGuruge) [inline build-config/resources/Ballerina.toml]: Shall we add `aws` as well? I think there's an existing label `vendor/aws`.

File: gradle.properties [hunk 15]
```diff
@@ -0,0 +1,13 @@
+group=io.ballerinax
+version=0.1.0
```
Comment (ThisaruGuruge) [inline gradle.properties]:
```suggestion
version=0.1.0-SNAPSHOT
```

File: gradle.properties [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+
+puppycrawlCheckstyleVersion=10.12.1
+checkstyleToolVersion=10.12.1
+githubSpotbugsVersion=5.0.14
+githubJohnrengelmanShadowVersion=8.1.1
+underCouchDownloadVersion=5.4.0
+researchgateReleaseVersion=2.8.0
+testngVersion=7.6.1
+
+ballerinaLangVersion=2201.8.0-20230830-220400-8a7556d8
```
Comment (ThisaruGuruge) [inline gradle.properties]: Let's use a released Ballerina version here.

File: native/src/main/java/io/ballerina/stdlib/java/jdbc/nativeimpl/CallProcessor.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+ * This class holds the utility methods involved with executing the call statements.
+ *
+ * @since 0.1.0
+ */
+public class CallProcessor { 
+    private CallProcessor() { 
+    }
+    
+    public static Object nativeCall(Environment env, BObject client, BObject paramSQLString, BArray recordTypes) {
+        return io.ballerina.stdlib.sql.nativeimpl.CallProcessor.nativeCall(env, client, paramSQLString,
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/stdlib/java/jdbc/nativeimpl/CallProcessor.java]: Can't we import the static method and use the method name here without the qualifiers?

File: native/src/main/java/io/ballerina/stdlib/java/jdbc/nativeimpl/ClientProcessor.java [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+public class ClientProcessor {
+
+    public static Object createClient(BObject client, BMap<BString, Object> clientConfig,
+                                      BMap<BString, Object> globalPool) {
+        String url = clientConfig.getStringValue(Constants.ClientConfiguration.URL).getValue();
+        if (!isJdbcUrlValid(url)) {
+            return ErrorGenerator.getSQLApplicationError("Invalid JDBC URL: " + url);
+        }
+        BString userVal = clientConfig.getStringValue(Constants.ClientConfiguration.USER);
+        String user = userVal == null ? null : userVal.getValue();
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/stdlib/java/jdbc/nativeimpl/ClientProcessor.java]: Can't we define a separate function to retrieve the values from the `clientConfig` while handling the null? The logic seems to be repetitive here.

File: native/src/main/java/io/ballerina/stdlib/java/jdbc/nativeimpl/ExecuteProcessor.java [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+
+import io.ballerina.runtime.api.Environment;
+import io.ballerina.runtime.api.values.BArray;
+import io.ballerina.runtime.api.values.BObject;
+import io.ballerina.stdlib.sql.parameterprocessor.DefaultStatementParameterProcessor;
+
+/**
+ * This class contains methods for executing SQL queries.
+ *
+ * @since 1.0.0
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/stdlib/java/jdbc/nativeimpl/ExecuteProcessor.java]:
```suggestion
 * @since 0.1.0
```

**Review by ThisaruGuruge — COMMENTED** [review]
- Let's add the GitHub workflows for build and release. (Use the connector templates from the Ballerina library repo).
- Seems like the required checks on PRs are different from the actual ones. Shall we check on that?

File: .github/workflows/ci.yml
Comment (ThisaruGuruge) [inline .github/workflows/ci.yml:1]: Shall we use the workflow templates for the connectors?

File: .github/workflows/daily-build.yml
Comment (ThisaruGuruge) [inline .github/workflows/daily-build.yml:1]: Ditto

File: native/src/main/java/io/ballerina/stdlib/java/jdbc/Constants.java [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.ballerina.stdlib.java.jdbc;
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/stdlib/java/jdbc/Constants.java]: Is this package name correct? Shouldn't it be:

```suggestion
package io.ballerina.lib.aws.redshift;
```

File: native/src/main/java/io/ballerina/stdlib/java/jdbc/nativeimpl/CallProcessor.java (hunk 17 above)
Comment (aashikam) [reply native/src/main/java/io/ballerina/stdlib/java/jdbc/nativeimpl/CallProcessor.java]: The classname itself is also CallProcessor. I tried but it didn't work?

Comment (ThisaruGuruge) [reply native/src/main/java/io/ballerina/stdlib/java/jdbc/nativeimpl/CallProcessor.java]: I see. Then this is okay,

**Review by ThisaruGuruge — APPROVED** [review]

**Pull request merged**